### ISSUE TITLE
Reports duplicate file, renamed with ~, but then doesn't install the new file

### ISSUE THREAD

**Issue by leafo** [luarocks/luarocks#1276]
I Install a module (that was previously installed as a different version):

```
(master) ~/sites/streak.club > luarocks install --local --lua-version=5.1  mailgun
Installing https://luarocks.org/mailgun-1.2.0-1.src.rock

mailgun 1.2.0-1 depends on lua >= 5.1 (5.1-1 provided by VM)
mailgun 1.2.0-1 depends on lpeg (1.0.1-1 installed)
mailgun 1.2.0-1 depends on luasocket (3.0rc1-1 installed)
mailgun 1.2.0-1 depends on lua-cjson (2.1.0-1 installed)
mailgun 1.2.0-1 depends on luasec (0.6-1 installed)
Warning: /home/leafo/.luarocks/share/lua/5.1/mailgun/init.lua is not tracked by this installation of LuaRocks. Moving it to /home/leafo/.luarocks/share/lua/5.1/mailgun/init.lua~
mailgun 1.2.0-1 is now installed in /home/leafo/.luarocks (license: MIT)

Checking stability of dependencies in the absence of
mailgun dev-1...

Removing mailgun dev-1...
Removal successful.
```

Note the line `Warning: /home/leafo/.luarocks/share/lua/5.1/mailgun/init.lua is not tracked by this installation of LuaRocks. Moving it to /home/leafo/.luarocks/share/lua/5.1/mailgun/init.lua~`. 

`mailgun/init.lua` is part of a regular installation, as seen in the rockspec: https://luarocks.org/manifests/leafo/mailgun-1.2.0-1.rockspec I also know that this file, `init.lua`, has always been a part of the module, since I wrote it.

Then when I go look at what's in the LuaRocks tree:

```
leafo ~/.luarocks > find . | grep mailgun
./lib/luarocks/rocks-5.1/mailgun
./lib/luarocks/rocks-5.1/mailgun/1.2.0-1
./lib/luarocks/rocks-5.1/mailgun/1.2.0-1/mailgun-1.2.0-1.rockspec
./lib/luarocks/rocks-5.1/mailgun/1.2.0-1/doc
./lib/luarocks/rocks-5.1/mailgun/1.2.0-1/doc/README.md
./lib/luarocks/rocks-5.1/mailgun/1.2.0-1/rock_manifest
./share/lua/5.1/mailgun
./share/lua/5.1/mailgun/util.lua
./share/lua/5.1/mailgun/init.lua~
```

Note that `./share/lua/5.1/mailgun/init.lua~` is there is no `./share/lua/5.1/mailgun/init.lua`. So it installed a broken version of the module!

Running `luarocks install --local --lua-version=5.1  mailgun` will reinstall it again and then the required file will be there.

I'm having a hard time reproducing this, but I have now encountered this issue twice when updating some dependencies on two separate servers I run. (related to the downtime for luarocks.org yesterday). I have a feeling it's related to having a tree originally created for an older version of LuaRocks. Sadly I don't have a copy of the manifest files to see what exactly was there (but I am now keeping my .luarocks folder under git so I can see what changes if the error pops up again)

Minimally though, I would expect it not to install a broken version of the module, perhaps something is happening out of order with the removal -> installation. Maybe it would be worth having a verification step after installation to ensure all the files it expects are actually there.

----
```
leafo ~ > luarocks --version
/usr/bin/luarocks 3.4.0
LuaRocks main command-line interface
```

**Comment by hishamhm** [comment]
> I'm having a hard time reproducing this, but I have now encountered this issue twice when updating some dependencies on two separate servers I run.

I could not reproduce the problem exactly (`init.lua` installed every time here), but I managed to reproduce the error message, by installing an old version using LuaRocks 2.x and upgrading with LuaRocks 3.x (and doing similar tweaks to the repo's manifest by hand to achieve a similar result). 

The behavior change is related to files called `foo.init` (they registered in the manifest as `foo` in LR2 — which [caused other problems](https://github.com/luarocks/luarocks/commit/65c417e0ecda55f44c691df032163a8c08f0b52a) — and `foo.init` in LR3). I added a backwards-compatibility check to try to fallback to the LR2 manifest format, and the `init.lua~` file doesn't seem to be left behind anymore. I believe that (a636669) will fix the issue.

> Minimally though, I would expect it not to install a broken version of the module, perhaps something is happening out of order with the removal -> installation. Maybe it would be worth having a verification step after installation to ensure all the files it expects are actually there.

I'm adding a double-check pass to verify that all files from the rock's manifest are present in the deploy locations at the end of deployment (9fb04a9). That way, if anything of this sort happens again, at least you'll get an error message and failure `$?` to tell you something went wrong.

**Comment by hishamhm** [comment]
@leafo Merged a fix! I'm not sure #1289 fixed the issue completely since I could not fully reproduce your particular error state, but it did catch and fix other inconsistencies, and it should at the very least detect it and not let it fail silently (though there's a good chance it fixed it as well). I think we can close this one and reopen if it happens again. Thanks for reporting and sorry for the mess up!
